internal/client/client.go

```diff
@@ -188,6 +188,77 @@ func (c *CSAPI) SetRoomAccountData(t *testing.T, roomID string, eventType string
 }
 
+// GetAllPushRules fetches all configured push rules for a user from the homeserver.
+// Push rules are returned as a parsed gjson result
+//
+// Example of printing the IDs of all underride rules of the current user:
+//
+//	allPushRules := c.GetAllPushRules(t)
+//	globalUnderridePushRules := allPushRules.Get("global").Get("underride").Array()
+//
+//	for index, rule := range globalUnderridePushRules {
+//	  fmt.Printf("This rule's ID is: %s\n", rule.Get("rule_id").Str)
+//	}
+//
+// Push rules are returned in the same order received from the homeserver.
+func (c *CSAPI) GetAllPushRules(t *testing.T) gjson.Result {
+	t.Helper()
+
+	// We have to supply an empty string to the end of this path in order to generate a trailing slash.
+	// See https://github.com/matrix-org/matrix-spec/issues/457
+	res := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "pushrules", ""})
+	pushRulesBytes := ParseJSON(t, res)
+	return gjson.ParseBytes(pushRulesBytes)
+}
+
+// GetPushRule queries the contents of a client's push rule by scope, kind and rule ID.
+// A parsed gjson result is returned. Fails the test if the query to server returns a non-2xx status code.
+//
+// Example of checking that a global underride rule contains the expected actions:
+//
+//	containsDisplayNameRule := c.GetPushRule(t, "global", "underride", ".m.rule.contains_display_name")
+//	must.MatchGJSON(
+//	  t,
+//	  containsDisplayNameRule,
+//	  match.JSONKeyEqual("actions", []interface{}{
+//	    "notify",
+//	    map[string]interface{}{"set_tweak": "sound", "value": "default"},
+//	    map[string]interface{}{"set_tweak": "highlight"},
+//	  }),
+//	)
+func (c *CSAPI) GetPushRule(t *testing.T, scope string, kind string, ruleID string) gjson.Result {
+	t.Helper()
+
+	res := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "pushrules", scope, kind, ruleID})
+	pushRuleBytes := ParseJSON(t, res)
+	return gjson.ParseBytes(pushRuleBytes)
+}
+
+// SetPushRule creates a new push rule on the user, or modifies an existing one.
+// `before` and `after` parameters can be provided, which if set will map to the
+// `before` and `after` query parameters on the set push rules endpoint:
+// https://spec.matrix.org/v1.5/client-server-api/#get_matrixclientv3pushrules.
+//
+// Example of setting a push rule with ID 'com.example.rule2' that must come after 'com.example.rule1':
+//
+//	c.SetPushRule(t, "global", "underride", "com.example.rule2", map[string]interface{}{
+//	  "actions": []string{"dont_notify"},
+//	}, nil, "com.example.rule1")
+func (c *CSAPI) SetPushRule(t *testing.T, scope string, kind string, ruleID string, body map[string]interface{}, before *string, after *string) *http.Response {
+	t.Helper()
+
+	// If the `before` or `after` arguments have been provided, construct same-named query parameters
+	queryParams := url.Values{}
+	if before != nil {
+		queryParams.Add("before", *before)
+	}
+	if after != nil {
+		queryParams.Add("after", *after)
+	}
```

Comment:
Was slightly surprised that `before` and `after` are passed as pointers. I guess the point is to allow to construct requests of the form `...?before=`? (one param provided as empty string, another omitted).

(I note that the spec doesn't mandate that `rule_id`s should be nonempty strings, so I suppose `before` and `after` could be take to be possibly empty too.)

Reply:
Hmm, I did this as I figured passing `nil` to mean "omit the parameter" would be more natural than passing `""` to mean the same thing. But I suppose it does allow callers to pass `""` and thus send up an empty string value.

Which do you think is more intuitive? Whichever way we choose, I should probably include a clarifying line in the docstring.

Reply:
Looking around `client.go`, I don't see pointers being used as optional parameters anywhere else. So I think I'll revert to non-pointers for consistency (and to avoid people accidentally passing `""` when they meant `nil` :).

Reply:
Don't have any strong opinions (unusually); I'd try to be consistent with whatever complement already does and whatever's considered idomatic go.

Reply:
Sounds good! I'm going to go ahead with an empty string then - it's still relatively easy to use and prevents a potential `""` footgun.